### PR TITLE
crimson/osd: bring support for client blocklisting

### PR DESCRIPTION
**Pull request by rzarzynski**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/pg.cc
Comment (athanatos) [inline src/crimson/osd/pg.cc:136]: This is going to iterate over object contexts for other PGs as well.  The PG needs a set of its own objects with live watches -- there shouldn't in general be all that many.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/pg.cc
Comment (rzarzynski) [reply src/crimson/osd/pg.cc:136]: How about skipping a watch on PG mismatch?

File: src/common/intrusive_lru.h
Comment (rzarzynski) [inline src/common/intrusive_lru.h:172]: The previous version missed assuring that the `lru` isn't `nullptr`. Fixed.

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
@athanatos: would you mind another look?

File: src/common/intrusive_lru.h
Comment (athanatos) [reply src/common/intrusive_lru.h:173]: Why does this need a WITH_SEASTAR ifdef?

File: src/crimson/osd/pg.cc
Comment (athanatos) [inline src/crimson/osd/pg.cc:135]: [this](

File: src/crimson/osd/pg.cc
Comment (athanatos) [reply src/crimson/osd/pg.cc:135]: Iterating over all live object contexts on the shard to find the very small subset with a live watcher for this pg still strikes me as inefficient and complicated.

File: src/crimson/osd/watch.h
Comment (athanatos) [inline src/crimson/osd/watch.h:97]: ?

**Comment by athanatos** [comment]
Incidentally, if you re-request review it'll pop back up on my github query for PRs waiting on a review from me.

**Comment by Matan-B** [comment]
Ping

**Comment by athanatos** [comment]
Looks like comments haven't been addressed?

File: src/common/intrusive_lru.h
Comment (rzarzynski) [reply src/common/intrusive_lru.h:173]: The idea was to emphasize it doesn't pose a threat to the classical world, so -- despite affecting the `common` -- the big QA testing is not a must.

File: src/crimson/osd/pg.cc
Comment (rzarzynski) [reply src/crimson/osd/pg.cc:135]: I see. Perhaps we could flip the loops.

File: src/crimson/osd/pg.cc
Comment (rzarzynski) [reply src/crimson/osd/pg.cc:135]: Thanks!

File: src/crimson/osd/watch.h
Comment (rzarzynski) [reply src/crimson/osd/watch.h:97]: Junk. Fixing.

File: src/crimson/osd/pg.cc
Comment (rzarzynski) [reply src/crimson/osd/pg.cc:135]: Nope, we don't have a separate registry of watchers. The main difference with the classical is it uses a per-PG (not per-shard) registry of object contexts:

```cpp
void PrimaryLogPG::check_blocklisted_watchers()
{
  dout(20) << "PrimaryLogPG::check_blocklisted_watchers for pg " << get_pgid() << dendl;
  pair<hobject_t, ObjectContextRef> i;
  while (object_contexts.get_next(i.first, &i))
    check_blocklisted_obc_watchers(i.second);
}
```

Duplicating the information looks even worse. HOWEVER:

```cpp
  using lru_set_t = boost::intrusive::set<
    base_t,
    lru_set_option_t,
    boost::intrusive::key_of_value<VToKWrapped>
    >;
  lru_set_t lru_set;
```

Maybe, by changing the key derivation function, we could have OBCs sorted / prefixed by PG-derived part.

File: src/common/intrusive_lru.h
Comment (rzarzynski) [reply src/common/intrusive_lru.h:172]: Removed.

**Comment by rzarzynski** [comment]
The pushed revisions addresses just the minors; the iteration thing requires further discussion.

File: src/crimson/osd/pg.cc
Comment (athanatos) [reply src/crimson/osd/pg.cc]: Probably not, tbh.  We can just let each PG have its own.

Comment (rzarzynski) [reply src/crimson/osd/pg.cc]: https://github.com/ceph/ceph/pull/49729

File: src/crimson/osd/pg.cc
Comment (Matan-B) [inline src/crimson/osd/pg.cc:152]: `for_each_obc` traverses through `obc_lru` which includes only cached obc (up to `crimson_osd_obc_lru_size`).
If we need to `for_each` all the existing obc, the ones which are stored in disc should also be taken into account.
This may cause excessive disc IO's just for checking blocklisted watchers.

**Review by Matan-B — CHANGES_REQUESTED** [review]

File: src/crimson/osd/pg.cc
Comment (Matan-B) [reply src/crimson/osd/pg.cc:152]: Classic corresponding implementation uses `while (object_contexts.get_next`).
The key difference is that classic's `SharedLRU` can return objects even after they have been evicted from cache using `weak_refs` map.

See header comment:
```
/// SharedLRU does its best to cache objects. It not only tracks the objects
/// in its LRU cache with strong references, it also tracks objects with
/// weak_ptr even if the cache does not hold any strong references to them. so
/// that it can return the objects after they are evicted, as long as they've
/// ever been cached and have not been destroyed yet.
```
In Crimson we use `intrusive_lru` which AFAIU doesn't support weak refs (yet).
```
  void evict() {
    while (!unreferenced_list.empty() &&
	   lru_set.size() > lru_target_size) {
      auto &b = unreferenced_list.front();
      assert(!b.lru);
      unreferenced_list.pop_front();
      lru_set.erase_and_dispose(
	lru_set.iterator_to(b),
	[](auto *p) { delete p; }
      );
    }
  }
```

Comment (rzarzynski) [reply src/crimson/osd/pg.cc:152]: Working on it. I'm afraid that missing the _registry_ behavior might be far more problematic that just the client blocking per se. I wonder about a situation in which multiple `ObjectContext` for the same object are used the same time.

Comment (rzarzynski) [reply src/crimson/osd/pg.cc:152]: We're evicting only from the `unreferenced_list`. An instance is being put on it when the last smart pointer is taken out:

```cpp
template <typename Config>
void intrusive_ptr_release(intrusive_lru_base<Config> *p) {
  assert(p);
  assert(p->use_count > 0);
  --p->use_count;
  if (p->use_count == 0) {
    p->lru->unreferenced(*p);
  }
}
```

So, we should be good -- it's a registry, not only a cache.

Comment (Matan-B) [reply src/crimson/osd/pg.cc:152]: > So, we should be good -- it's a registry, not only a cache.

Yes, assuming we keep referencing relevant obcs.

I pushed #49779 for clarification.

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
@rzarzynski, Let's get this merged for `rbd_lock_and_fence` to pass in suite runs.

**Comment by rzarzynski** [comment]
@Matan-B: rebased. Let's finish this PR!

**Review by Matan-B — APPROVED** [review]

**Comment by rzarzynski** [comment]
```
The following tests FAILED:
	  2 - setup-venv-for-mgr (Failed)
	  4 - run-tox-mgr (Not Run)
	 34 - run-rbd-unit-tests-127.sh (Timeout)
```

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
Once make check is up again, this PR can be merged based on:

https://pulpito.ceph.com/matan-2023-04-30_16:46:54-crimson-rados-wip-matanb-crimson-only-testing-30.4-distro-crimson-smithi/
- 13/19 passed.
- Dead jobs: 
`7258784` - infra
`7258785` - infra
- Failed, known:
`7258780` - https://tracker.ceph.com/issues/59165
`7258781` - New test (tracked #50457)
`7258792` - rbd_lock_and_fence
`7258796` - rbd_python_api_tests_old_format

**Comment by rzarzynski** [comment]
```
The following tests FAILED:
	  2 - setup-venv-for-mgr (Failed)
	  4 - run-tox-mgr (Not Run)
```

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
jenkins test make check

**Pull request merged**
